test/unit-tests/mobile/android-device-discovery.ts

```diff
@@ -0,0 +1,271 @@
+					future.return(device);
+				});
+
+				androidDeviceDiscovery.checkForDevices().wait();
+				let output = `List of devices attached${EOL}`;
+				mockStdoutEmitter.emit('data', output);
+				let lostDevice = future.wait();
+				assert.deepEqual(lostDevice.deviceInfo.identifier, androidDeviceIdentifier);
+				assert.deepEqual(lostDevice.status, androidDeviceStatus);
+
+				androidDeviceDiscovery.on("deviceLost", (device: Mobile.IDevice) => {
+					throw new Error("Should not report device as removed next time after it has been already reported.");
+				});
+				androidDeviceDiscovery.checkForDevices().wait();
+				mockStdoutEmitter.emit('data', output);
+			});
+
+			it("reports it as removed and added after that next time when called and device's status is changed", () => {
+				let deviceLostFuture = new Future<any>();
+				let deviceFoundFuture = new Future<any>();
+				androidDeviceDiscovery.on("deviceLost", (device: Mobile.IDevice) => {
+					_.remove(devicesFound, d => d.deviceInfo.identifier === device.deviceInfo.identifier);
+					deviceLostFuture.return(device);
+				});
+
+				androidDeviceDiscovery.on("deviceFound", (device: Mobile.IDevice) => {
+					devicesFound.push(device);
+					deviceFoundFuture.return(device);
+				});
+
+				androidDeviceDiscovery.checkForDevices().wait();
+				let output = `List of devices attached${EOL}${androidDeviceIdentifier}	unauthorized${EOL}${EOL}`;
+				mockStdoutEmitter.emit('data', output);
+				let lostDevice = deviceLostFuture.wait();
+				assert.deepEqual(lostDevice.deviceInfo.identifier, androidDeviceIdentifier);
+				assert.deepEqual(lostDevice.status, androidDeviceStatus);
+
+				deviceFoundFuture.wait();
+				assert.isTrue(devicesFound.length === 1, "We should have found ONE device.");
+				assert.deepEqual(devicesFound[0].deviceInfo.identifier, androidDeviceIdentifier);
+				assert.deepEqual(devicesFound[0].status, "unauthorized");
+
+				// Verify the device will not be reported as found next time when adb returns the same output:
+				// In case it is reported, an error will be raised - Future resolved more than once for deviceFoundFuture
+				mockStdoutEmitter.emit('data', output);
+				assert.isTrue(devicesFound.length === 1, "We should have found ONE device.");
+			});
+		});
+
+		it("throws error when adb writes on stderr", () => {
+			androidDeviceDiscovery.on("deviceFound", (device: Mobile.IDevice) => {
+				throw new Error("Devices should not be found.");
+			});
+			let error = new Error("ADB Error");
+			try {
+				androidDeviceDiscovery.checkForDevices().wait();
+				mockStderrEmitter.emit('data', error);
+			} catch (err) {
+				assert.deepEqual(err, error);
+			}
+		});
+
+		it("throws error when adb's child process throws error", () => {
+			androidDeviceDiscovery.on("deviceFound", (device: Mobile.IDevice) => {
+				throw new Error("Devices should not be found.");
+			});
+			let error = new Error("ADB Error");
+			try {
```

Comment:
I prefer to keep it this way and verify the error. With `assert.throws` some other error may occur, but the test will still pass.